### PR TITLE
Apr 2923 organisations addresses data

### PR DESCRIPTION
**Pull request by raviudari01**

### CODE REVIEW

File: src/SFA.DAS.ApplyService.Application/Organisations/UpdateOrganisation/UpdateOrganisationHandler.cs [hunk 1]
```diff
@@ -39,5 +47,25 @@ private async Task<Organisation> UpdateOrganisationIfExists(UpdateOrganisationRe
 
             return null;
         }
+
+        private async Task<OrganisationAddresses> UpdateOrganisationAddressesIfExists(Organisation request)
+        {
```
Comment (DevenKShah) [inline src/SFA.DAS.ApplyService.Application/Organisations/UpdateOrganisation/UpdateOrganisationHandler.cs:52]: DRY

File: src/SFA.DAS.ApplyService.Application/Organisations/CreateOrganisation/CreateOrganisationHandler.cs [hunk 2]
```diff
@@ -25,12 +27,16 @@ public async Task<Organisation> Handle(CreateOrganisationRequest request, Cancel
 
             var result = await UpdateOrganisationIfExists(request) ?? await CreateNewOrganisation(request);
```
Comment (DevenKShah) [inline src/SFA.DAS.ApplyService.Application/Organisations/CreateOrganisation/CreateOrganisationHandler.cs:28]: Why are we updating in create when there is a separate update handler. This needs sorting out. Find the usage of both the handlers and see if you can either keep only one handler (upsert) or avoid updating in create handler so consumer determines whether to update or create and calls respective handler. I prefer the former approach of having a single handler.

**Review by DevenKShah — CHANGES_REQUESTED** [review]

File: src/SFA.DAS.ApplyService.Application/Organisations/UpdateOrganisation/UpdateOrganisationHandler.cs (hunk 1 above)
Comment (raviudari01) [reply src/SFA.DAS.ApplyService.Application/Organisations/UpdateOrganisation/UpdateOrganisationHandler.cs]: UpdateOrganisationHandler is deleted, it is not needed

File: src/SFA.DAS.ApplyService.Application/Organisations/CreateOrganisation/CreateOrganisationHandler.cs (hunk 2 above)
Comment (raviudari01) [reply src/SFA.DAS.ApplyService.Application/Organisations/CreateOrganisation/CreateOrganisationHandler.cs:28]: UpdateOrganisationHandler is deleted and CreateOrganisationHandler is renamed to ManageOrganisationHandler

File: src/SFA.DAS.ApplyService.Data/OrganisationAddressesRepository.cs [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+        public async Task<OrganisationAddresses> GetOrganisationAddressesByOrganisationId(Guid organisationId)
+        {
+            using (var connection = _dbConnectionHelper.GetDatabaseConnection())
+            {
+                var sql =
+                    @"SELECT *
+                    FROM OrganisationAddresses 
+                    WHERE OrganisationId = @OrganisationId";
+
+                var orgAddresses = await connection.QueryFirstOrDefaultAsync<OrganisationAddresses>(sql, new { organisationId });
```
[the remark this answers is not in the document]
Comment (raviudari01) [reply src/SFA.DAS.ApplyService.Data/OrganisationAddressesRepository.cs]: Made as list

File: src/SFA.DAS.ApplyService.Application.UnitTests/Handlers/OrganisationsHandlerTests/CreateOrganisationHandlerTests.cs [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+        [SetUp]
+        public void TestSetup()
+        {
+            _organisationRepository = new Mock<IOrganisationRepository>();
+            _organisationAddressesRepository = new Mock<IOrganisationAddressesRepository>();
+            _handler = new ManageOrganisationHandler(_organisationRepository.Object, _organisationAddressesRepository.Object);
+        }
+
+        [Test]
+        public async Task CreateOrganisationHandler_creates_new_Organization_if_it_not_exists()
```
Comment (DevenKShah) [inline src/SFA.DAS.ApplyService.Application.UnitTests/Handlers/OrganisationsHandlerTests/CreateOrganisationHandlerTests.cs:29]: Test names to follow standard naming convention

File: src/SFA.DAS.ApplyService.Application/Organisations/ManageOrganisation/ManageOrganisationHandler.cs [hunk 5]
```diff
@@ -72,6 +72,47 @@ private async Task<Organisation> UpdateOrganisationIfExists(CreateOrganisationRe
...
+
+            organisationAddresses.Id = await _organisationAddressesRepository.CreateOrganisationAddresses(organisationAddresses);
+            return organisationAddresses;
+        }
+
+        private async Task<OrganisationAddresses> UpdateOrganisationAddressesIfExists(Organisation request)
+        {
+            var existingOrganisationAddresses = await _organisationAddressesRepository.GetOrganisationAddressesByOrganisationId(request.Id);
+
+            if (existingOrganisationAddresses != null && existingOrganisationAddresses.Count > 0)
```
Comment (DevenKShah) [inline src/SFA.DAS.ApplyService.Application/Organisations/ManageOrganisation/ManageOrganisationHandler.cs:96]: Invert if, return early

File: src/SFA.DAS.ApplyService.Application/Organisations/ManageOrganisation/ManageOrganisationHandler.cs [hunk 6]
```diff
@@ -72,6 +72,47 @@ private async Task<Organisation> UpdateOrganisationIfExists(CreateOrganisationRe
...
+            {
+                var organisationAddressToUpdate = existingOrganisationAddresses.Find(a => a.AddressType == (int)OrganisationAddressType.LegalAddress);
```
Comment (DevenKShah) [inline src/SFA.DAS.ApplyService.Application/Organisations/ManageOrganisation/ManageOrganisationHandler.cs:98]: why find why not FirstOrDefault?

File: src/SFA.DAS.ApplyService.Application/Organisations/ManageOrganisation/ManageOrganisationHandler.cs (hunk 5 above)
Comment (raviudari01) [reply src/SFA.DAS.ApplyService.Application/Organisations/ManageOrganisation/ManageOrganisationHandler.cs]: Done

File: src/SFA.DAS.ApplyService.Application.UnitTests/Handlers/OrganisationsHandlerTests/CreateOrganisationHandlerTests.cs (hunk 4 above)
Comment (raviudari01) [reply src/SFA.DAS.ApplyService.Application.UnitTests/Handlers/OrganisationsHandlerTests/CreateOrganisationHandlerTests.cs]: Done

File: src/SFA.DAS.ApplyService.Application/Organisations/ManageOrganisation/ManageOrganisationHandler.cs (hunk 6 above)
Comment (raviudari01) [reply src/SFA.DAS.ApplyService.Application/Organisations/ManageOrganisation/ManageOrganisationHandler.cs]: Done

**Review by DevenKShah — APPROVED** [review]

**Pull request merged**